sim/src/mechanics/intersection.rs

```diff
@@ -838,8 +868,60 @@ impl IntersectionSimState {
         // If a case #1 could've started by now, then they would have. Since they didn't, they must
         // be blocked.
 
-        // TODO Make sure we can optimistically finish this turn before an approaching
-        // higher-priority vehicle wants to begin.
+        // Can we optimistically finish this turn before an approaching higher-priority vehicle
+        // wants to begin?
+
+        let turn = map.get_t(req.turn);
+        let expected_finish = now + turn.geom.length() / speed;
+        for (other_req, eta) in &self.state[&req.turn.parent].leader_eta {
+            if expected_finish < *eta {
+                continue;
+            }
+            let other_turn = map.get_t(other_req.turn);
+            if other_turn.conflicts_with(turn) {
+                // Now we need to prioritize between the two. First use the stop sign priority
+                let their_priority = sign.get_priority(other_req.turn, map);
+                let should_yield = if our_priority < their_priority {
+                    true
+                } else if our_priority > their_priority {
+                    false
+                } else {
+                    // Same priority by the stop sign, so check turn types
+                    let turn_type_ranks = if map.get_config().driving_side == DrivingSide::Right {
+                        vec![TurnType::Straight, TurnType::Right, TurnType::Left]
+                    } else {
+                        vec![TurnType::Straight, TurnType::Left, TurnType::Right]
+                    };
+
+                    // Other turn types (U-turns and pedestrian crossings) always lose. That
+                    // should probably be configurable
+                    let our_rank = turn_type_ranks
+                        .iter()
+                        .position(|x| *x == turn.turn_type)
+                        .unwrap_or(3);
+                    let their_rank = turn_type_ranks
+                        .iter()
+                        .position(|x| *x == other_turn.turn_type)
+                        .unwrap_or(3);
+                    // 0 is the highest rank here
+                    // If all the priorities are equal, we should go -- we're ready to start;
+                    // they're still approaching.
+                    our_rank > their_rank
```

Comment:
Very cool. You get to encode the driver's interpretations of the road rules and priority. Even let them take risks (like following a leader into an intersection, hoping traffic will keep flowing until you get out).

Reply:
Discrete events mean these decisions only happen at a few moments, and it's important to "wake everyone up" anytime there's a potential decision to make. And the complication this PR is exposing is the difference between people at the intersection already waiting, and those approaching it. But in principle, yes, one place to decide this ordering.